topdown/strings.go

```diff
@@ -10,10 +10,180 @@ import (
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			strs = append(strs, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, builtins.NewOperandTypeErr(2, b, "set", "array")
+	}
+
+	var prefixes []string
+	switch b := b.(type) {
+	case *ast.Array:
+		err := b.Iter(func(x *ast.Term) error {
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			prefixes = append(prefixes, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	case ast.Set:
+		err := b.Iter(func(x *ast.Term) error {
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			prefixes = append(prefixes, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, builtins.NewOperandTypeErr(2, b, "set", "array")
+	}
+
+	trie := patricia.NewTrie()
+	for i := 0; i < len(strs); i++ {
+		trie.Insert([]byte(strs[i]), true)
+	}
+
+	for i := 0; i < len(prefixes); i++ {
+		if trie.MatchSubtree([]byte(prefixes[i])) {
+			return ast.Boolean(true), nil
+		}
+	}
+
+	return ast.Boolean(false), nil
+}
+
+func builtinAnySuffixMatchesAnyString(a, b ast.Value) (ast.Value, error) {
+
+	var strs []string
+	switch a := a.(type) {
+	case *ast.Array:
+		err := a.Iter(func(x *ast.Term) error {
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			strs = append(strs, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	case ast.Set:
+		err := a.Iter(func(x *ast.Term) error {
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			strs = append(strs, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, builtins.NewOperandTypeErr(2, b, "set", "array")
+	}
+
+	var suffixes []string
+	switch b := b.(type) {
+	case *ast.Array:
+		err := b.Iter(func(x *ast.Term) error {
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			suffixes = append(suffixes, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	case ast.Set:
+		err := b.Iter(func(x *ast.Term) error {
+			s, ok := x.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandElementErr(2, b, x.Value, "string")
+			}
+			suffixes = append(suffixes, string(s))
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, builtins.NewOperandTypeErr(2, b, "set", "array")
+	}
+
+	trie := patricia.NewTrie()
+	for i := 0; i < len(strs); i++ {
+		trie.Insert([]byte(reverseString(strs[i])), true)
+	}
+
+	for i := 0; i < len(suffixes); i++ {
+		if trie.MatchSubtree([]byte(reverseString(suffixes[i]))) {
```

Comment:
Is the string reversal the only thing different between these two functions?

Reply:
Yep. I've extracted the common functionality out now, while keeping the argument parsing separate. Happy to deduplicate further, but also didn't want to overcomplicate it.

Reply:
Great! I think that's enough. DRY isn't in fashion these days, or so I've heard... so this should be a good compromise :)